File: web/src/garden/strings.ts
/**
 * Все сербские надписи сада в одном месте.
 *
 * Игра говорит по-сербски — в этом её смысл, — но каждая надпись несёт русское
 * пояснение: без него новичок A1 не поймёт, на что нажимает. Письмо кириллица;
 * если решим перейти на латиницу, правится только этот файл.
 */

export interface Phrase {
  sr: string;
  ru: string;
}

export const GARDEN = {
  title: { sr: 'Башта Читавука', ru: 'Сад Читавука' },
  coins: { sr: 'цветни динари', ru: 'цветочные динары' },
  plant: { sr: 'Посади', ru: 'посадить' },
  water: { sr: 'Залиј', ru: 'полить' },
  shop: { sr: 'Продавница семена', ru: 'магазин семян' },
  empty: { sr: 'Празна леја', ru: 'пустая грядка' },
  board: { sr: 'Најбољи баштовани', ru: 'лучшие садоводы' },
  neighbours: { sr: 'Комшије', ru: 'соседи' },
  helpNeighbour: { sr: 'Залиј комшији', ru: 'полить соседу' },
  speed: { sr: 'Брзина раста', ru: 'скорость роста' },
  earnings: { sr: 'Данашња зарада', ru: 'заработок за сегодня' },
  bloomed: { sr: 'Процветало', ru: 'распустилось' },
  myName: { sr: 'Име баштована', ru: 'имя садовода' },
  openGarden: { sr: 'Отвори башту комшијама', ru: 'показывать сад другим' },
  save: { sr: 'Сачувај', ru: 'сохранить' },
  practise: { sr: 'Вежбај', ru: 'позаниматься' },
  find: { sr: 'Нађи баштована', ru: 'найти садовода' },
  growing: { sr: 'Гаји', ru: 'выращивает' },
  nobody: { sr: 'Нема никога', ru: 'никого не нашлось' },

  can: { sr: 'Канта за воду', ru: 'лейка' },
  fill: { sr: 'Захвати воду', ru: 'набрать воды' },
  riverAwake: { sr: 'Река тече', ru: 'река проснулась' },
  riverDry: { sr: 'Река спава', ru: 'река спит' },
  cut: { sr: 'Убери цвет', ru: 'срезать цветок' },
  herbarium: { sr: 'Хербаријум', ru: 'гербарий' },
  task: { sr: 'Задатак дана', ru: 'задание дня' },
  done: { sr: 'Урађено', ru: 'сделано' },
  rain: { sr: 'Киша', ru: 'дождь' },
  clear: { sr: 'Сунчано', ru: 'ясно' },
  night: { sr: 'Ноћ', ru: 'ночь' },

  enter: { sr: 'Уђи у кућу', ru: 'войти в дом' },
  leave: { sr: 'Изађи', ru: 'выйти во двор' },
  home: { sr: 'Код куће', ru: 'дома' },
  radio: { sr: 'Радио', ru: 'радиоприёмник' },
  radioOn: { sr: 'Упали', ru: 'включить' },
  radioOff: { sr: 'Угаси', ru: 'выключить' },
  tuning: { sr: 'Тражи станицу', ru: 'ищет станцию' },
  noSignal: { sr: 'Не хвата', ru: 'станция не отвечает' },
  furnish: { sr: 'Уреди кућу', ru: 'обставить дом' },
  notebook: { sr: 'Свеска речи', ru: 'тетрадь со словами' },
  buy: { sr: 'Купи', ru: 'купить' },
  owned: { sr: 'Купљено', ru: 'куплено' },
} as const;

/**
 * Подписи предметов мира. Ради них сад и затевался: сербское слово должно
 * попадаться там, где человек и так смотрит, а не в списке слов.
 */
export const WORLD: Record<string, Phrase> = {
  house: { sr: 'кућа', ru: 'дом' },
  tree: { sr: 'дрво', ru: 'дерево' },
  fir: { sr: 'јела', ru: 'ель' },
  fountain: { sr: 'фонтана', ru: 'фонтан' },
  campfire: { sr: 'ватра', ru: 'костёр' },
  river: { sr: 'река', ru: 'река' },
  bed: { sr: 'леја', ru: 'грядка' },
  fence: { sr: 'ограда', ru: 'забор' },
  stall: { sr: 'продавница', ru: 'магазин' },
  duck: { sr: 'патка', ru: 'утка' },
  flowers: { sr: 'цвеће', ru: 'цветы' },
  bushes: { sr: 'жбуње', ru: 'кусты' },
  pots: { sr: 'саксије', ru: 'горшки' },
  sign: { sr: 'путоказ', ru: 'указатель' },
};

/** Задание дня по-сербски. Цель подставляется числом. */
export function taskPhrase(kind: string, target: number): Phrase {
  switch (kind) {
    case 'water':
      return { sr: `Залиј леје: ${target}`, ru: `полей грядки: ${target}` };
    case 'plant':
      return { sr: `Посади цвет: ${target}`, ru: `посади цветок: ${target}` };
    case 'cut':
      return { sr: `Убери цвет: ${target}`, ru: `срежь цветок: ${target}` };
    case 'help':
      return { sr: `Залиј комшији: ${target}`, ru: `полей соседу: ${target}` };
    default:
      return { sr: 'Задатак дана', ru: 'задание дня' };
  }
}

/** Стадии роста. Последняя — распустившийся цветок. */
export const STAGES: Phrase[] = [
  { sr: 'семе', ru: 'семя' },
  { sr: 'клица', ru: 'росток' },
  { sr: 'стабљика', ru: 'стебель' },
  { sr: 'пупољак', ru: 'бутон' },
  { sr: 'цвет', ru: 'цветок' },
];

/**
 * Какая доля цветка видна на этой стадии.
 *
 * Цветок растёт обрезкой снизу вверх: у подсолнуха и ромашки стебель длинный, а
 * головка сверху, поэтому он буквально поднимается из земли, и отдельные
 * картинки на каждую стадию не нужны.
 */
export function stageHeight(stage: number, stages: number): number {
  if (stages <= 1) return 100;
  const steps = [12, 34, 62, 84, 100];
  const index = Math.min(Math.max(stage, 0), Math.min(stages, steps.length) - 1);
  return steps[index] ?? 100;
}

export function plantImage(species: string, stage?: number): string {
  if (stage !== undefined) {
    return `/img/garden/world/plant_${species}_${Math.min(4, Math.max(0, stage))}.webp`;
  }
  return `/img/garden/plant_${species}.webp`;
}

/** Смещение семени в атласе `garden_seeds.webp`. */
export function seedOffset(species: string, catalog: readonly { id: string }[]): number {
  const index = catalog.findIndex((item) => item.id === species);
  return index < 0 ? 0 : index;
}

export function coinWord(count: number): string {
  const tail = Math.abs(count) % 100;
  if (tail >= 11 && tail <= 14) return 'динаров';
  switch (tail % 10) {
    case 1:
      return 'динар';
    case 2:
    case 3:
    case 4:
      return 'динара';
    default:
      return 'динаров';
  }
}

File: frontend/lib/garden/strings.dart
/// Все сербские надписи сада в одном месте.
///
/// Копия `web/src/garden/strings.ts`: сад один, и подписи в приложении обязаны
/// совпадать с сайтом слово в слово. Игра говорит по-сербски — в этом её
/// смысл, — но каждая надпись несёт русское пояснение: без него новичок A1 не
/// поймёт, на что нажимает.
library;

class Phrase {
  const Phrase(this.sr, this.ru);

  final String sr;
  final String ru;
}

class Garden {
  const Garden._();

  static const title = Phrase('Башта Читавука', 'Сад Читавука');
  static const coins = Phrase('цветни динари', 'цветочные динары');
  static const plant = Phrase('Посади', 'посадить');
  static const water = Phrase('Залиј', 'полить');
  static const shop = Phrase('Продавница семена', 'магазин семян');
  static const empty = Phrase('Празна леја', 'пустая грядка');
  static const board = Phrase('Најбољи баштовани', 'лучшие садоводы');
  static const neighbours = Phrase('Комшије', 'соседи');
  static const helpNeighbour = Phrase('Залиј комшији', 'полить соседу');
  static const speed = Phrase('Брзина раста', 'скорость роста');
  static const earnings = Phrase('Данашња зарада', 'заработок за сегодня');
  static const bloomed = Phrase('Процветало', 'распустилось');
  static const myName = Phrase('Име баштована', 'имя садовода');
  static const openGarden = Phrase('Отвори башту комшијама', 'показывать сад другим');
  static const save = Phrase('Сачувај', 'сохранить');
  static const practise = Phrase('Вежбај', 'позаниматься');
  static const find = Phrase('Нађи баштована', 'найти садовода');
  static const growing = Phrase('Гаји', 'выращивает');
  static const nobody = Phrase('Нема никога', 'никого не нашлось');

  static const can = Phrase('Канта за воду', 'лейка');
  static const fill = Phrase('Захвати воду', 'набрать воды');
  static const riverAwake = Phrase('Река тече', 'река проснулась');
  static const riverDry = Phrase('Река спава', 'река спит');
  static const cut = Phrase('Убери цвет', 'срезать цветок');
  static const herbarium = Phrase('Хербаријум', 'гербарий');
  static const task = Phrase('Задатак дана', 'задание дня');
  static const done = Phrase('Урађено', 'сделано');
  static const rain = Phrase('Киша', 'дождь');
  static const clear = Phrase('Сунчано', 'ясно');
  static const night = Phrase('Ноћ', 'ночь');

  static const enter = Phrase('Уђи у кућу', 'войти в дом');
  static const leave = Phrase('Изађи', 'выйти во двор');
  static const home = Phrase('Код куће', 'дома');
  static const radio = Phrase('Радио', 'радиоприёмник');
  static const radioOn = Phrase('Упали', 'включить');
  static const radioOff = Phrase('Угаси', 'выключить');
  static const furnish = Phrase('Уреди кућу', 'обставить дом');
  static const notebook = Phrase('Свеска речи', 'тетрадь со словами');
  static const buy = Phrase('Купи', 'купить');
  static const owned = Phrase('Купљено', 'куплено');
  static const fridgeOpen = Phrase('Шта има у фрижидеру', 'что в холодильнике');
  static const letters = Phrase('Ћирилица и латиница', 'кириллица и латиница');
  static const fireOn = Phrase('Ватра гори', 'огонь горит');
  static const fireOff = Phrase('Ватра се угасила', 'огонь погас');
  static const toBooks = Phrase('Отвори књиге', 'открыть материалы');
}

/// Подписи предметов двора. Ради них сад и затевался: сербское слово должно
/// попадаться там, где человек и так смотрит, а не в списке слов.
const Map<String, Phrase> worldWords = {
  'house': Phrase('кућа', 'дом'),
  'tree': Phrase('дрво', 'дерево'),
  'fir': Phrase('јела', 'ель'),
  'fountain': Phrase('фонтана', 'фонтан'),
  'campfire': Phrase('ватра', 'костёр'),
  'river': Phrase('река', 'река'),
  'bed': Phrase('леја', 'грядка'),
  'fence': Phrase('ограда', 'забор'),
  'stall': Phrase('продавница', 'магазин'),
  'duck': Phrase('патка', 'утка'),
  'flowers': Phrase('цвеће', 'цветы'),
  'bushes': Phrase('жбуње', 'кусты'),
  'pots': Phrase('саксије', 'горшки'),
  'sign': Phrase('путоказ', 'указатель'),
};

/// Вещи в квартире. Ради них дом и открывается.
const Map<String, Phrase> houseWords = {
  'door': Phrase('врата', 'дверь'),
  'window': Phrase('прозор', 'окно'),
  'picture': Phrase('слика', 'картина'),
  'blackboard': Phrase('табла', 'доска'),
  'bed': Phrase('кревет', 'кровать'),
  'nightstand': Phrase('ноћни сточић', 'тумбочка'),
  'radio': Phrase('радио', 'радиоприёмник'),
  'tvstand': Phrase('комода', 'комод'),
  'tv': Phrase('телевизор', 'телевизор'),
  'fireplace': Phrase('камин', 'камин'),
  'sofa': Phrase('кауч', 'диван'),
  'rug': Phrase('тепих', 'ковёр'),
  'shelf': Phrase('полица', 'полка'),
  'desk': Phrase('радни сто', 'письменный стол'),
  'notebook': Phrase('свеска', 'тетрадь'),
  'lamp': Phrase('лампа', 'лампа'),
  'cat': Phrase('мачка', 'кошка'),
  'pot': Phrase('саксија', 'горшок с цветком'),
  'kitchen': Phrase('кухиња', 'кухня'),
  'fridge': Phrase('фрижидер', 'холодильник'),
  'table': Phrase('сто', 'стол'),
  'chair-left': Phrase('столица', 'стул'),
  'chair-right': Phrase('столица', 'стул'),
};

/// Что лежит в холодильнике: первая сербская еда, которую видишь в магазине.
const List<Phrase> fridgeWords = [
  Phrase('млеко', 'молоко'),
  Phrase('сир', 'сыр'),
  Phrase('јаја', 'яйца'),
  Phrase('хлеб', 'хлеб'),
  Phrase('јогурт', 'йогурт'),
  Phrase('кајмак', 'каймак'),
  Phrase('паприка', 'перец'),
  Phrase('сок', 'сок'),
];

/// Сербский алфавит на доске: кириллица и латиница буква в букву.
const List<List<String>> alphabet = [
  ['А а', 'A a'], ['Б б', 'B b'], ['В в', 'V v'], ['Г г', 'G g'], ['Д д', 'D d'],
  ['Ђ ђ', 'Đ đ'], ['Е е', 'E e'], ['Ж ж', 'Ž ž'], ['З з', 'Z z'], ['И и', 'I i'],
  ['Ј ј', 'J j'], ['К к', 'K k'], ['Л л', 'L l'], ['Љ љ', 'Lj lj'], ['М м', 'M m'],
  ['Н н', 'N n'], ['Њ њ', 'Nj nj'], ['О о', 'O o'], ['П п', 'P p'], ['Р р', 'R r'],
  ['С с', 'S s'], ['Т т', 'T t'], ['Ћ ћ', 'Ć ć'], ['У у', 'U u'], ['Ф ф', 'F f'],
  ['Х х', 'H h'], ['Ц ц', 'C c'], ['Ч ч', 'Č č'], ['Џ џ', 'Dž dž'], ['Ш ш', 'Š š'],
];

/// Стадии роста. Последняя — распустившийся цветок.
const List<Phrase> growthStages = [
  Phrase('семе', 'семя'),
  Phrase('клица', 'росток'),
  Phrase('стабљика', 'стебель'),
  Phrase('пупољак', 'бутон'),
  Phrase('цвет', 'цветок'),
];

/// Задание дня по-сербски. Цель подставляется числом.
Phrase taskPhrase(String kind, int target) => switch (kind) {
      'water' => Phrase('Залиј леје: $target', 'полей грядки: $target'),
      'plant' => Phrase('Посади цвет: $target', 'посади цветок: $target'),
      'cut' => Phrase('Убери цвет: $target', 'срежь цветок: $target'),
      'help' => Phrase('Залиј комшији: $target', 'полей соседу: $target'),
      _ => Garden.task,
    };

/// «динар», «динара», «динаров» — как в русском, по последней цифре.
String coinWord(int count) {
  final tail = count.abs() % 100;
  if (tail >= 11 && tail <= 14) return 'динаров';
  return switch (tail % 10) {
    1 => 'динар',
    2 || 3 || 4 => 'динара',
    _ => 'динаров',
  };
}
